client/pd_service_discovery.go

```diff
@@ -386,29 +442,30 @@ func (c *pdServiceDiscovery) initClusterID() error {
 	return nil
 }
 
-func (c *pdServiceDiscovery) updateServiceMode() {
+func (c *pdServiceDiscovery) checkServiceModeChanged() error {
 	leaderAddr := c.getLeaderAddr()
-	if len(leaderAddr) > 0 {
-		clusterInfo, err := c.getClusterInfo(c.ctx, leaderAddr, c.option.timeout)
-		// If the method is not supported, we set it to pd mode.
-		if err != nil {
+	if len(leaderAddr) == 0 {
+		return errors.New("no leader found")
+	}
+
+	clusterInfo, err := c.getClusterInfo(c.ctx, leaderAddr, c.option.timeout)
+	if err != nil {
+		if strings.Contains(err.Error(), "Unimplemented") {
+			// If the method is not supported, we set it to pd mode.
 			// TODO: it's a hack way to solve the compatibility issue.
 			// we need to remove this after all maintained version supports the method.
-			if strings.Contains(err.Error(), "Unimplemented") {
-				c.serviceModeUpdateCb(pdpb.ServiceMode_PD_SVC_MODE)
-			} else {
-				log.Warn("[pd] failed to get cluster info for the leader", zap.String("leader-addr", leaderAddr), errs.ZapError(err))
-			}
-			return
+			c.serviceModeUpdateCb(pdpb.ServiceMode_PD_SVC_MODE)
+			return nil
 		}
-		c.serviceModeUpdateCb(clusterInfo.ServiceModes[0])
-	} else {
-		log.Warn("[pd] no leader found")
+		return err
 	}
+
+	c.serviceModeUpdateCb(clusterInfo.ServiceModes[0])
```

Comment:
Do we need to validate the `ServiceModes` filed to make sure it has at least one element?

Reply:
if clusterInfo == nil || len(clusterInfo.ServiceModes) == 0 {
		return errors.WithStack(errNoServiceModeReturned)
	}
	c.serviceModeUpdateCb(clusterInfo.ServiceModes[0])